### PR TITLE
Studio: Add ability to connect and disconnect a site

### PR DESCRIPTION
**Pull request by katinthehatsite**
Closes: https://github.com/Automattic/dotcom-forge/issues/9354
Closes: https://github.com/Automattic/dotcom-forge/issues/9357

## Proposed Changes

This PR adds the ability to connect and disconnect a site and ensures that the data is persistent per user.  In this PR, we are moving away from relying on the local state and instead save the data about the connected sites to the `appdata-v1.json`.

## Testing Instructions

**Testing per user**

* Log into Studio with account for user A 
* Navigate to `Sync` tab on the site A 
* Click on `Connect site`
* Observe that you see a site connected 
* Switch to the site B
* Observe that you see no connected site on the site B
* Go back to site A to the `Sync` tab
* Click on `Connect site`
* Select a site that both has a staging site and a production site
* Connect that site
* Observe that the two sites get added together
* Click on `Disconnect` button
* Confirm that you see the warning about disconnecting the site 
* Confirm that the warning contains `Don't ask again` option
* Disconnect the site
* Observe that the two sites get disconnected together
* Generally switch between tabs and sites and confirm that the data persists

----
* Log out and log back in with user B
* Observe that you see no sites connected for user B (as expected)
* Connect some sites to user B
* Log out
* Log in with user A
* Confirm that no connected sites from user B are present for user A 

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/components/sync-connected-sites.tsx [hunk 1]
```diff
@@ -17,43 +17,81 @@ interface ConnectedSiteSection {
...
+	}, [ connectedSites ] );
+
+	const handleDisconnectSite = async ( sectionId: number, sectionName?: string ) => {
+		const dontShowDisconnectWarning = localStorage.getItem( 'dontShowDisconnectWarning' );
+		if ( ! dontShowDisconnectWarning ) {
+			const CANCEL_BUTTON_INDEX = 1;
+			const DISCONNECT_BUTTON_INDEX = 0;
+
+			const disconnectMessage = sectionName
+				? __( 'Disconnect' ) + ' ' + sectionName
```
Comment (sejas) [inline src/components/sync-connected-sites.tsx]: Let's use `sprintf` instead of concat two strings. That will improve the experience for RTL languages.

```suggestion
				? sprintf( __( 'Disconnect %s' ), sectionName )
```

File: src/components/sync-connected-sites.tsx [hunk 2]
```diff
@@ -17,43 +17,81 @@ interface ConnectedSiteSection {
...
+			const { response, checkboxChecked } = await getIpcApi().showMessageBox( {
+				type: 'info',
+				message: disconnectMessage,
+				detail: __(
+					'Your WordPress.com site will not be affected by disconnecting it from Studio.'
+				),
+				buttons: [ __( 'Disconnect' ), __( 'Cancel' ) ],
+				cancelId: CANCEL_BUTTON_INDEX,
+				checkboxLabel: __( "Don't ask again" ),
+				checkboxChecked: false,
```
Comment (sejas) [inline src/components/sync-connected-sites.tsx]: I think we don't need to pass the checkboxChecked or the type info attributes.

```suggestion
				message: disconnectMessage,
				detail: __(
					'Your WordPress.com site will not be affected by disconnecting it from Studio.'
				),
				buttons: [ __( 'Disconnect' ), __( 'Cancel' ) ],
				cancelId: CANCEL_BUTTON_INDEX,
				checkboxLabel: __( "Don't ask again" ),
```

File: src/hooks/use-site-sync-management.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+	return {
+		connectedSites,
+		loadConnectedSites,
+		connectSite,
+		disconnectSite,
+	} as const;
+};
+
+export { useSiteSyncManagement };
```
Comment (sejas) [inline src/hooks/use-site-sync-management.ts]: Curious about why not just adding export to the left of the function declaration on line 7

File: src/ipc-handlers.ts [hunk 4]
```diff
@@ -216,6 +217,99 @@ export async function updateSite(
...
+	// Connect staging sites if any
+	if ( stagingSites && stagingSites.length > 0 ) {
+		stagingSites.forEach( ( stagingSite ) => {
+			addConnection( {
+				...stagingSite,
+				isStaging: true,
+				stagingSiteIds: [],
+			} );
+		} );
+	}
```
Comment (sejas) [inline src/ipc-handlers.ts]: What do you think about removing the auxiliar function `addConnection` ?

Could be written something like:

```ts
	const isAlreadyConnected = connections.some(
		( conn ) => conn.id === siteToConnect.id && conn.localSiteId === localSiteId
	);

	if ( ! isAlreadyConnected ) {
		connections.push( { ...siteToConnect, localSiteId } );
	}

	// Connect staging sites if any
	if ( stagingSites && stagingSites.length > 0 ) {
		stagingSites.forEach( ( stagingSite ) => {
			connections.push( { ...stagingSite, localSiteId } );
		} );
	}
```

File: src/ipc-handlers.ts [hunk 5]
```diff
@@ -216,6 +217,99 @@ export async function updateSite(
 	return mergeSiteDetailsWithRunningDetails( userData.sites );
 }
 
+export async function connectWpcomSite(
+	event: IpcMainInvokeEvent,
+	site: SyncSite,
```
Comment (sejas) [inline src/ipc-handlers.ts]: Not a strong opinion, but maybe we could pass directly an array of sites, and calculate the related staging sites before calling this function. In a similar way how `disconnectWpcomSite` currently works.

File: src/ipc-handlers.ts [hunk 6]
```diff
@@ -216,6 +217,99 @@ export async function updateSite(
...
+
+	if ( ! currentUserId ) {
+		throw new Error( 'User not authenticated' );
+	}
+
+	const connections = userData.connectedWpcomSites?.[ currentUserId ] || [];
+
+	const updatedConnections = connections.filter(
+		( conn ) => ! ( siteIds.includes( conn.id ) && conn.localSiteId === localSiteId )
+	);
```
Comment (sejas) [inline src/ipc-handlers.ts:269]: Is this filter removing the connections that are matching siteIds and localSiteId?
Maybe we could rename the variable name?

File: src/ipc-handlers.ts [hunk 7]
```diff
@@ -216,6 +217,99 @@ export async function updateSite(
...
+): Promise< SyncSite[] > {
+	const userData = await loadUserData();
+	const currentUserId = userData.authToken?.id;
+
+	if ( ! currentUserId ) {
+		return [];
+	}
+
+	return (
+		userData?.connectedWpcomSites?.[ currentUserId ]?.filter(
```
Comment (sejas) [inline src/ipc-handlers.ts]: We don't need this question mark. We already accessed this object on line 300.

```suggestion
		userData.connectedWpcomSites?.[ currentUserId ]?.filter(
```

File: src/hooks/use-fetch-wpcom-sites.tsx [hunk 8]
```diff
@@ -116,10 +118,15 @@ export function useSyncSites() {
...
-		isFetching: isFetchingSites.current,
-	};
-}
+	// Map syncSites to reflect whether they are already connected
+	const syncSitesWithConnectionStatus = syncSites.map( ( site ) => ( {
+		...site,
+		syncSupport: connectedSites.some( ( connectedSite ) => connectedSite.id === site.id )
+			? 'already-connected'
+			: site.syncSupport,
+	} ) );
```
Comment (sejas) [inline src/hooks/use-fetch-wpcom-sites.tsx]: This could be wrapped with useMemo

File: src/components/content-tab-sync.tsx [hunk 9]
```diff
@@ -162,23 +163,33 @@ function NoAuthSyncTab() {
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 export function ContentTabSync( { selectedSite }: { selectedSite: SiteDetails } ) {
 	const { __ } = useI18n();
-	const { syncSites, connectedSites, setConnectedSites, isFetching } = useSyncSites();
+	const { connectedSites, connectSite, disconnectSite } = useSiteSyncManagement();
+	const { syncSites, isFetching } = useFetchWpComSites( connectedSites );
```
Comment (sejas) [inline src/components/content-tab-sync.tsx]: We are already calling `useFetchWpComSites` inside `useSiteSyncManagement`. Could we remove the fetch from this component in favor of the call inside the hook?

https://github.com/Automattic/studio/blob/de20256c1fdfcc066d3988e25ac18a23c0073011/src/hooks/use-site-sync-management.ts#L10

Something like: 

```suggestion
	const { connectedSites, connectSite, disconnectSite, syncSites, isFetching } = useSiteSyncManagement();
```

File: src/hooks/use-site-sync-management.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
+import { useState, useEffect, useCallback } from 'react';
+import { getIpcApi } from '../lib/get-ipc-api';
+import { useAuth } from './use-auth';
+import { SyncSite, useFetchWpComSites } from './use-fetch-wpcom-sites';
+import { useSiteDetails } from './use-site-details';
+
+const useSiteSyncManagement = () => {
+	const [ connectedSites, setConnectedSites ] = useState< SyncSite[] >( [] );
+	const { isAuthenticated } = useAuth();
+	const { syncSites } = useFetchWpComSites( connectedSites );
```
Comment (sejas) [inline src/hooks/use-site-sync-management.ts]: As mentioned before, let's keep this fetch and return `syncSites` and `isFetching` from this hook.

File: src/hooks/use-site-sync-management.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	}, [ isAuthenticated, loadConnectedSites ] );
+
+	const connectSite = useCallback(
+		async ( site: SyncSite ) => {
+			if ( ! localSiteId ) {
+				return;
+			}
+			try {
+				const stagingSites = site.stagingSiteIds
+					.map( ( id ) => syncSites?.find( ( s ) => s.id === id ) )
```
Comment (sejas) [inline src/hooks/use-site-sync-management.ts]: Do we need this question mark?

```suggestion
					.map( ( id ) => syncSites.find( ( s ) => s.id === id ) )
```

File: src/hooks/use-site-sync-management.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+					.filter( ( s ): s is SyncSite => s !== undefined );
```
Comment (sejas) [inline src/hooks/use-site-sync-management.ts]: Let's remove this type declaration. It seems is not needed.

```suggestion
					.filter( ( s ) => s !== undefined );
```

**Review by sejas — COMMENTED** [review]
@katinthehatsite Great work!, I've tested the feature and I'm able to connect and disconnect sites. It works smoothly.
I see them updated in my appData.
It works when login in and login out, keeping the previous connected sites.

We may need to test some edge cases when we receive from the backend some missing sites that are already connected.

I've left some comments and suggestions. Feel free to push back in anything you disagree with.
I would also like to add some tests around the feature, in this PR or create an issue for that.

I've noticed a quick blinking when we change tabs, but we can improve that once we create a Context :D


https://github.com/user-attachments/assets/00ba3970-3d60-410e-b872-8ba6be1a644e

**Comment by katinthehatsite** [comment]
> I would also like to add some tests around the feature, in this PR or create an issue for that.

Sounds good, I have added a separate issue for tests: https://github.com/Automattic/dotcom-forge/issues/951

File: src/components/sync-connected-sites.tsx (hunk 1 above)
Comment (katinthehatsite) [reply src/components/sync-connected-sites.tsx]: sounds good, I applied the suggestion https://github.com/Automattic/studio/pull/610/commits/90de8b01bcb0ae6e8dc245638c35741cf15f37b0

File: src/components/sync-connected-sites.tsx (hunk 2 above)
Comment (katinthehatsite) [reply src/components/sync-connected-sites.tsx]: I removed them here: https://github.com/Automattic/studio/pull/610/commits/3c99d84e2a28ba8e55853449da7ae0bf830ff357

The reason why I passed `checkboxChecked` was because I was not sure if it should be checked or unchecked by default so I added them there in case we want to default it to checked.

File: src/hooks/use-site-sync-management.ts (hunk 3 above)
Comment (katinthehatsite) [reply src/hooks/use-site-sync-management.ts]: I think this is just the leftover of my refactoring, I was moving many things around and adjusting them. I do not even like this approach 😂 

Fixed in https://github.com/Automattic/studio/pull/610/commits/041411d7351f66ad4748992184afc40d1c28e184

File: src/ipc-handlers.ts (hunk 7 above)
Comment (katinthehatsite) [reply src/ipc-handlers.ts]: Fixed in https://github.com/Automattic/studio/pull/610/commits/28b3e1e9bee7e92e83254d8e5b93c08d75a6b372

File: src/hooks/use-fetch-wpcom-sites.tsx (hunk 8 above)
Comment (katinthehatsite) [reply src/hooks/use-fetch-wpcom-sites.tsx]: Done in https://github.com/Automattic/studio/pull/610/commits/c36c9530a1b2a065bce69c4039d0dc8fc0d73d3c

File: src/components/content-tab-sync.tsx (hunk 9 above)
Comment (katinthehatsite) [reply src/components/content-tab-sync.tsx]: Updated in https://github.com/Automattic/studio/pull/610/commits/91efc5acbb687e0e33a892b75764508e13549f13

File: src/hooks/use-site-sync-management.ts (hunk 10 above)
Comment (katinthehatsite) [reply src/hooks/use-site-sync-management.ts]: Updated in https://github.com/Automattic/studio/pull/610/commits/91efc5acbb687e0e33a892b75764508e13549f13

File: src/hooks/use-site-sync-management.ts (hunk 12 above)
Comment (katinthehatsite) [reply src/hooks/use-site-sync-management.ts]: Donre in https://github.com/Automattic/studio/pull/610/commits/91efc5acbb687e0e33a892b75764508e13549f13

File: src/hooks/use-site-sync-management.ts (hunk 11 above)
Comment (katinthehatsite) [reply src/hooks/use-site-sync-management.ts]: Adjusted in https://github.com/Automattic/studio/pull/610/commits/bef0f2fde09a8ae6740fcdbcb5d29705403f83f3

File: src/ipc-handlers.ts (hunk 4 above)
Comment (katinthehatsite) [reply src/ipc-handlers.ts]: No strong preference here - I guess we can drop it as we do not really need to make another check for the staging sites connection.

I make some changes here: https://github.com/Automattic/studio/pull/610/commits/16fe606f7f1c45c991f8b06070ef7fc748723950

File: src/hooks/use-site-sync-management.ts (hunk 12 above)
Comment (katinthehatsite) [reply src/hooks/use-site-sync-management.ts]: Actually I had to undo this in https://github.com/Automattic/studio/pull/610/commits/c120e3ceee5c625cc26561baafa5be02e86eba6e - I was not able to make the types pass without this one.

@sejas Did it work for you in your testing without this type declaration?

File: src/ipc-handlers.ts (hunk 6 above)
Comment (katinthehatsite) [reply src/ipc-handlers.ts:269]: Yes that is right. @sejas. I do not have a strong preference - perhaps `remainingConnections` or something like that? Do you have any specific suggestions in mind?

File: src/ipc-handlers.ts (hunk 5 above)
Comment (katinthehatsite) [reply src/ipc-handlers.ts]: Made changes in https://github.com/Automattic/studio/pull/610/commits/05e94314ad5c87144aa010fe921cde49a83be942 to address this. I think it makes sense to have this so that there is a consistency between these two functions

**Comment by katinthehatsite** [comment]
@sejas I responded to your suggestions - let me know what you think! We can add a separate issue for creating a context

File: src/hooks/use-site-sync-management.ts (hunk 12 above)
Comment (sejas) [reply src/hooks/use-site-sync-management.ts]: @katinthehatsite  ha!, you were right, the tests failed if we didn't force the types.
It's weird that the IDE didn't raise that error but the tests did, maybe we are using different versions of eslint.

Here is the error I received:

```
 FAIL  src/components/tests/content-tab-sync.test.tsx
  ● Test suite failed to run

    src/hooks/use-site-sync-management.ts:43:6 - error TS2322: Type '{ syncSupport: SyncSupport; id: number; localSiteId: string; name: string; url: string; isStaging: boolean; stagingSiteIds: number[]; } | undefined' is not assignable to type 'SyncSite'.
      Type 'undefined' is not assignable to type 'SyncSite'.

    43      ...site.stagingSiteIds
            ~~~~~~~~~~~~~~~~~~~~~~
    44       .map( ( id ) => syncSites.find( ( s ) => s.id === id ) )
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    45       .filter( ( s ) => s !== undefined ),
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 FAIL  src/components/tests/site-content-tabs.test.tsx
  ● Test suite failed to run

    src/hooks/use-site-sync-management.ts:43:6 - error TS2322: Type '{ syncSupport: SyncSupport; id: number; localSiteId: string; name: string; url: string; isStaging: boolean; stagingSiteIds: number[]; } | undefined' is not assignable to type 'SyncSite'.
      Type 'undefined' is not assignable to type 'SyncSite'.

    43      ...site.stagingSiteIds
            ~~~~~~~~~~~~~~~~~~~~~~
    44       .map( ( id ) => syncSites.find( ( s ) => s.id === id ) )
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    45       .filter( ( s ) => s !== undefined ),
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

After expending more time, and trying to see if we can remove the forced type there, I think this particular case was good use of it. 👍 

Although, I also found another approach that works good.
It seems the problem in the types was because .find can return undefined, and then we filtered them, but the types were inconsistent. With flatMap we return an empty array that flatMap removes if is empty when it flattens the array.

Here is my update. Let me know what you think:

https://github.com/Automattic/studio/pull/610/commits/394a13b3f8f4982ceda62a98345bd8bef6731732

**Review by sejas — APPROVED** [review]
Thanks for addressing my suggestions! It looks great.
I tested it by connecting/disconnecting the sites, and log in/out the user, and confirm I see the connected sites for a given user.

https://github.com/user-attachments/assets/cf08d0c8-8bda-422c-9eed-e1e99781bc91

I've found a corner case, where we should clean the connected sites when we remove a local site. That was not described in the original scope of the issue so I created a new one:

- https://github.com/Automattic/dotcom-forge/issues/9548
